### ISSUE TITLE
Batchnorm hangs in affine application

### ISSUE THREAD

**Issue by rhewett** [distdl/distdl#182]
For some reason batchnorm is not allowing re-use of the broadcast layer for the affine application.

A temporary patch will be applied but we'll need to get to the root of this.

**Comment by jmerizia** [comment]
I did some more testing in #181, and isolated the issue to the Broadcast layer. This issue might be related. Is there a snippet of code / error message / specific environment where this one seems to occur?

**Comment by rhewett** [comment]
The patch is merged, you can see it here: https://github.com/distdl/distdl/commit/00ca001f6f932c7ae0dc11ab36a0fefe764082e0

Based on the sizes and dtypes of the variables, there should be no issues re-using the broadcast layer.  Broadcast does not store any internal state and does not use extra buffers.

In the case I encountered, it was a 2x2 convolution on square data.  (256x256?  I'll verify later.)

And it is weird because it is already used twice before it hits the affine block (once for mean and once for var).  But as soon as we use it for a third time (i.e., for gamma or for beta if we use the "second" layer for gamma) ranks 1-3 do not exit.  

This is for both OpenMPI (on the clusters) and for mpiche (locally).  So, I'm not sure if it is the same issue as #181 or not.

**Comment by rhewett** [comment]
This appears to also impact the transpose layers too.  Once cannot scatter two tensors with the same layer without a deadlock.  Running within a `torch.no_grad()` region does not change the result.

**Comment by rhewett** [comment]
More on this...

For the transpose, it shows up in the scatter pattern.  Here, any processor not in the root will have the same empty tensor as input so of course the check will fail.  So somehow we need to either make the global tensor shape always available when it is created (wrap the torch tensor creation routines), force a flag when we know the tensor shape resets, or add a collective to the check.

I want to avoid collectives across the tensors because of cost but we may be forced to for stability purposes until a better solution is found.
